Yet another fix to stuck at home

hhaim opened the pull request:
Signed-off-by: Hanoch Haim <hhaim@cisco.com>

## Breaking Change:

<!-- What is breaking and why we have to break it. Remove this section only if it was NOT a breaking change. -->

## Description:
Yet another fix to stuck at home


**Related issue (if applicable):** fixes #<home-assistant issue number goes here>

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#<home-assistant.io PR number goes here>

## Example entry for `configuration.yaml` (if applicable):
```yaml

```

## Checklist:
  - [ x] The code change is tested and works locally.
  - [ ] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [ x] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [ ] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [ ] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [ ] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [ ] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard/__init__.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard/__init__.py#L23

hhaim committed:
yet another fix into the buggy device tracker

Signed-off-by: Hanoch Haim <hhaim@cisco.com>

balloob commented:
You don't have to open new PRs to make changes. Just push it to your branch.

Also please describe how your fix fixes things, and what was broken.

hhaim commented:
I'm not sure I've followed all the complex logic of the code. What was clear to me that the *aging* should be always called.  I think that there are still many issues in the code and I did not  fix all of them (only what fix my issue). The fact that there were so many "stuck at home" issues  just corroborate that it is  a complex modue.

For my case  I've started to fix stuff and found the root issue only at the end. 
Let me start  with the root issue:

**The issue**:
* My *custom* component for dnsmasq call 
  see(mac=xxx,location=STATE_HOME) when the device ask for IP from dhcp server. 
by *mistake* I've provided location=STATE_HOME (without that there is no issue)
* The ping didn't work due to firewall issue 
* This device was stuck forever @ HOME state. 

**The reason**:

This code is called for all the device to find stall 
```
        for device in self.devices.values():
            if (device.track and (device.state == STATE_HOME)) and \
               device.stale(now):
                self.hass.async_create_task(device.async_update_ha_state(True))
```

When it finds a device that is in a stall condition (tract && @home && dtime>consider_home) it call this function 

```
    async def async_update(self):
   ..
        if self.location_name:
            self._state = self.location_name

```

But in my case the self.location_name is "home" so it didn't do anything ==> *stuck at home *

moving the stall() to be the first solve the issue

2. Moving the stall to be first won't change the stall for cases of location ```NOT_AT_HOME``` but I think we should do this too : something like that 

```
or device in self.devices.values():
            if (device.track and  device.stale(now)) :
                self.hass.async_create_task(device.async_update_ha_state(True))
```

Any location should be stall after timeout not just HOME (why would we want to stall on arbitrary location for infinite time??)
But this is a *different* PR 

3. More issues in the code 

3.1 
``self.last_seen`` should not be None in *any* case. Not  sure why it checks for this condition ..
There should be an assert in case it is None (it will simplify the code)

So to make sure we don't have this condition I've added this, but better to remove the None 

a workaround 
```
        if not self.last_seen:
            self.last_seen = dt_util.utcnow()
            return

```

and this one 
```
        return self.last_seen is None or \
            (now or dt_util.utcnow()) - self.last_seen > self.consider_home
```

3.2 

```
 self.last_update_home
```
This variable is redundant and can only cause bugs by having it-- can be replaced by a function that read the (self._state == HOME). I've removed it

hhaim commented:
Is there a reason to keep a PR open so long?
Could you say what is the reason?

balloob commented on homeassistant/components/device_tracker/__init__.py:
```diff
@@ -382,7 +382,7 @@ def async_update_stale(self, now: dt_util.dt.datetime):
         """
         for device in self.devices.values():
-            if (device.track and device.last_update_home) and \
+            if (device.track and (device.state == STATE_HOME)) and \
```
This won't work because we're currently conflating the concerns of location based trackers and network based trackers, meaning a location set to HOME by a location tracker should never be overriden to be marked as stale or last seen updated.

We have decoupled the two implementations so that future platforms will not have this issue. Older platforms will be slowly migrated to the new style.

hhaim commented:
@balloob Thanks for taking a look
It been a long time since I did this PR so I don’t remember.
Is this issue solved with the new code base?
I think that any change to HOME (whichever component) should be age after a time in all cases. Is it?

balloob commented:
We're fixing it by splitting the different concerns. It's still a work in progress, and depending on the platform you're using, it might not be fixed yet.
